[Team-19][AOS] [ Linus& 히데 Airbnb 첫번째주 수요일 PR ]  홈, 로그인, 검색 화면 UI 구현

HyungMinKang opened the pull request:
## 진행사항
-  스타일, 테마 정의
-  로그인 화면
    -  로그인 fragment 생성
    - 로그인 하지 않고 진행하기 버튼 클릭시 navigate 구현

- 홈 화면
    -  상단 검색바 구현 -> 클릭심 검색 화면으로 이동
    -  히어로 이미지 + 이미지 위에 뷰 처리 구현
    -  근처 여행지 리사이클러뷰 구현
    -  추천 여행지 리사이클러뷰 구현
    -  bottomNavigation + NavController 구현

- 검색 화면
   -   TextInputLayout+ editText로  검색바 구현
   -  TextWatcher를 통해  검색어가 없을시와 있을시 UI 변경 로직 구현
   
## 결과
![ezgif com-gif-maker (1)](https://user-images.githubusercontent.com/58967292/170199697-b2bbc4f5-007e-46d3-b8d8-d9604eb8fee8.gif)

## 추후 진행사항
- 리사이클러뷰를  중첩하는 방법으로 Custom Calendar 구현을 진행할 예정입니다. 

## 질문 
![ezgif com-gif-maker](https://user-images.githubusercontent.com/58967292/170199902-1424d7cf-370b-480d-925a-1a7cf4a9f333.gif)

``` Kotlin
   private fun addSearchListener(searchAdapter: SearchResultAdapter) {
        binding.etSearch.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {}

            override fun onTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {}

            override fun afterTextChanged(searchKey: Editable?) {
                if (searchKey.isNullOrEmpty()) {
                    displayNearDestination()
                } else {
                    displaySearchResultDestination()
                    val dummyList = makeDummySearchResultByKeyword(searchKey.toString())
                    searchAdapter.submitSearchResultDestinations(dummyList)
                }
            }
        })
    private fun displayNearDestination(){
        binding.clSearchNearTravelDestination.visibility= View.VISIBLE
        binding.rvSearchResultDestination.visibility= View.GONE
        binding.etlSearch.setEndIconDrawable(R.drawable.ic_search)
        binding.etlSearch.endIconMode= END_ICON_CUSTOM
        binding.etlSearch.setEndIconTintList(ColorStateList.valueOf(ContextCompat.getColor(requireContext(),R.color.pink)))
    }

    private fun displaySearchResultDestination(){
        binding.clSearchNearTravelDestination.visibility= View.GONE
        binding.rvSearchResultDestination.visibility= View.VISIBLE
        binding.etlSearch.setEndIconDrawable(R.drawable.ic_clear)
        binding.executePendingBindings()
        binding.etlSearch.endIconMode= END_ICON_CLEAR_TEXT
        binding.etlSearch.setEndIconTintList(ColorStateList.valueOf(ContextCompat.getColor(requireContext(),R.color.grey1)))
    }

```
textWatcher 에서 공백이 아니면 endIcon 의 drawable과 mode, tint 를 바꾸는 로직을 구성했는데, 이상하게도 mode(clearText), tint 등은 정상적으로 작동하는데 drawable 코드만 작동하지 않습니다. 위의 gif 에서 1글자를 입력하면 돋보기 모양의 아이콘이 x 아이콘으로 변경돼야 하는데 이 부분만 변경되지 않습니다. 혹시 어떤 문제인지 조언을 구해도 될까요?

kang-proto committed:
test

cherrylime69 committed:
Project : airbnb project 시작

HyungMinKang committed:
Merge pull request #1 from hwicode/android-develop

Project : airbnb project 시작

kang-proto committed:
feat: 스타일 테마 정의

- color 추가
- style에 글자 크기별 스타일 정의

cherrylime69 committed:
Merge pull request #3 from hwicode/feature/style

[Android]  히데& 라이너스  스타일 테마 정의

kang-proto committed:
feat: 로그인 화면 구현

- 로그인 fragment 생성
- 로그인 하지 않고 진행하기 버튼 클릭시 navigate 구현

cherrylime69 committed:
Merge pull request #5 from hwicode/feature/login

[Android] 히데 로그인 화면 구현

cherrylime69 committed:
design : 홈 화면 레이아웃 일부 구성
- 검색바 구성
- 히어로 이미지(이미지만) 구성
- Bottom navigation 구현

HyungMinKang committed:
Merge pull request #7 from hwicode/feature/main-bottomNavigation

Bottom Navigation 구현

kang-proto committed:
feat: home 화면 UI 마무리

- 더미 데이터로 근처여행지, 추천여행지 리사이클러뷰 부분 설정
- RecommendDestination, NearDestination 모델 정의
- Glide App 모듈 정의
- 히어로 이미지 위에 올라가는 뷰 구현
- ImageUrl을 뷰에 로딩하기 위한 BindingAdapter 정의
- 로그인 화면 둥근 버튼을 위해 drawble 파일 정의

cherrylime69 committed:
Merge pull request #9 from hwicode/feature/homeFragment

[Android][히데]  홈 화면 UI 완성

cherrylime69 committed:
feat : 검색 화면 일부 구현
- TextInput 을 통해 텍스트 넣도록 구현
- 검색어가 빈값이면 근처의 인기 여행지가 리사이클러뷰 보이도록 구현

HyungMinKang committed:
Merge pull request #11 from hwicode/feature/home-search

feat : 검색 화면 일부 구현

kang-proto committed:
feat: 검색 화면 완성

- 검색 textInputLayout에 textWatcher 추가
- 검색어 유무에 따라 UI 변경
- 검색어 결과 리사이클러뷰 , 추천 여행지 리사이클러뷰 추가
- bottomNavigation과 NavControllter 연결

cherrylime69 committed:
Merge pull request #13 from hwicode/feature/search-result

[Android][히데] 검색 화면 UI 완성

cherrylime69 committed:
refactor : RecyclerViewAdapter 를 ListAdapter 로 변경
- SearchResultAdapter 를 ListAdapter 로 변경

HyungMinKang committed:
Merge pull request #15 from hwicode/feature/search-result-refactor

RecyclerViewAdapter 를 ListAdapter 로 변경

HyungMinKang committed:
Merge pull request #16 from hwicode/android

첫번째 수요일 PR

cherrylime69 commented:
Linus : 반갑습니다 앨런. 미숙하지만 3주 동안 잘 부탁드리겠습니다! 주신 피드백은 최대한 고민하고 반영할 수 있도록 최선을 다하겠습니다 😁

sabgilhun commented on app/build.gradle:
```diff
@@ -0,0 +1,60 @@
+    implementation 'androidx.appcompat:appcompat:1.4.1'
+    implementation 'com.google.android.material:material:1.6.0'
+    implementation 'androidx.constraintlayout:constraintlayout:2.1.3'
+    implementation 'androidx.legacy:legacy-support-v4:1.0.0'
```
`androidx.legacy:legacy-support-v4`는 어떤 이유로 추가되었나요?

cherrylime69 replied:
실수로 추가되었습니다. 앞으로 커밋 하기 전에 한번 더 변경된 사항을 확인하도록 하겠습니다.

sabgilhun commented on app/src/main/java/com/example/airbnb/domain/model/NearDestination.kt:
```diff
@@ -0,0 +1,7 @@
+
+data class NearDestination(
+    val imageUrl: String,
+    val destination: String,
```
`destination`는 `NearDestination`의 이름을 뜻하는 property 맞나요? 

맞다면 그냥 `name`같은 네이밍은 어떤가요?

네이밍은 개인의 취향이기 때문에 반영하지 않으셔도 됩니다만, 

저 같은 경우는 property는 말그대로 그 class에서의 속성을 나타내야 한다고 생각하여 의견 드려봅니다.

cherrylime69 replied:
저희도 앨런 의견에 동의하여 name 으로 리네임하였습니다.

sabgilhun commented on app/src/main/java/com/example/airbnb/ui/MainActivity.kt:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    private fun showBottomNav() {
+        binding.bottomNavigationView.visibility = View.VISIBLE
```
Core KTX를 추가 하셨기 때문에 (`implementation 'androidx.core:core-ktx:1.7.0'`)

```kotlin
binding.bottomNavigationView.isVisible = true
```

위와같이 `View.isVisible`을 사용 가능합니다.

HyungMinKang replied:
isVisible로 수정하였습니다

sabgilhun commented on app/src/main/res/layout/item_search_result_destination.xml:
```diff
@@ -0,0 +1,49 @@
+        android:layout_marginEnd="16dp"
+        android:layout_marginBottom="20dp">
+
+        <androidx.cardview.widget.CardView
+            android:id="@+id/cv_search_result_destination"
+            android:layout_width="70dp"
+            android:layout_height="70dp"
+            app:cardCornerRadius="20dp"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toTopOf="parent">
+
+            <ImageButton
+                android:layout_width="match_parent"
+                android:layout_height="match_parent"
+                android:background="@color/grey_scale_off_white"
+                android:src="@drawable/ic_locationpin"
+                app:tint="@color/grey3" />
+
+        </androidx.cardview.widget.CardView>
```
`CardView`의 경우 기본적으로 elevation 값이 있어서, 그림자가 노출됩니다.

만약 elevation을 사용할 생각이라면, 그림자가 노출될 수 있게 `CardView` 주변 margin을 넣어 주어야 합니다.
지금은 여백이 없기 때문에 그림자가 잘려 보입니다.

elevation 적용하지 않을 생각이라면, `app:cardElevation="0dp"` 속성을 넣어 제거해주시면 됩니다.

HyungMinKang replied:
app:cardElevation="0dp"  속성을 넣어  elevation을 제거했습니다

sabgilhun commented on app/src/main/java/com/example/airbnb/ui/home/SearchFragment.kt:
```diff
@@ -0,0 +1,92 @@
+        binding.clSearchNearTravelDestination.visibility= View.GONE
+        binding.rvSearchResultDestination.visibility= View.VISIBLE
+        binding.etlSearch.setEndIconDrawable(R.drawable.ic_clear)
+        binding.executePendingBindings()
```
`executePendingBindings()`는 테스트 하다가 들어간 코드일까요?

맞다면 제거해도 무방해 보입니다 ~

HyungMinKang replied:
icon변경 여부가 반영이 안되서 테스트 하다가 들어간 코드인것 같습니다.  다음부터는 제출전에  수정하도록 주의하겠습니다.

sabgilhun reviewed (commented):
안녕하세요. Linus, 히데 3주간 잘 부탁드립니다.

코드가 전반적으로 깔끔한것 같네요. 좀더 깔끔한 코드를 위해 코드 포멧팅을 자주 해주시면 더 좋을것 같습니다.

### 질문 답변

`setEndIconDrawable()` 메서드 설명을 보니, 꼭 `setEndIconMode()` 메서드 호출 이후에 호출하라고 되어 있네요.

```kotlin
private fun displaySearchResultDestination(){
    binding.clSearchNearTravelDestination.visibility= View.GONE
    binding.rvSearchResultDestination.visibility= View.VISIBLE
    binding.etlSearch.endIconMode= END_ICON_CLEAR_TEXT
    binding.etlSearch.setEndIconDrawable(R.drawable.ic_clear)
    binding.etlSearch.setEndIconTintList(ColorStateList.valueOf(ContextCompat.getColor(requireContext(),R.color.grey1)))
}
```

위처럼 수정하니 정상동작 하는것 같습니다.

그런데, 이렇게 순서가 정해져 있는 API는 좋지 않아보이네요. (왜 이렇게 설계했는지...) 추후에도 계속 실수하기 쉬워 보이고요.

하나의 함수로 묶어 Util 함수를 만들고 프로젝트 내에서 사용할 수 있도록 해서 조금이나마 실수를 줄일 수 있지 않을까 합니다.

### 키보드
<img src="https://user-images.githubusercontent.com/38038123/170288907-55833ccb-84f3-4219-b356-d8af78fe8fe4.jpg" width="50%" height="50%" title="keyboard"/>

키보드 위로 `BottomNavigationView`가 올라오는게 약간 어색한것 같아요. `windowSoftInputMode` 이 속성으로 수정 가능해 보입니다~

sabgilhun commented on app/src/main/java/com/example/airbnb/ui/MainActivity.kt:
```diff
@@ -0,0 +1,43 @@
+            binding.bottomNavigationView.setupWithNavController(it)
+        }
+
+        navController?.addOnDestinationChangedListener { _, destination, _ ->
+            when (destination.id) {
+                R.id.loginFragment-> hideBottomNav()
+                else -> showBottomNav()
+            }
+        }
```
해당 코드는 MainActivity에서 현재 위에 떠 있는 Fragment에 따라 BottomNav를 띄울지 말지 컨트롤하는 코드로 보이는데요.

약간 아쉬운 점이 있는데요. 띄울지 말지를 MainActivity가 Fragment들을 하나 하나 확인하여 컨트롤하는 점입니다.

나중에는 Fragment의 수가 늘어나게 될거고, BottomNav말고도, AppBar등 더 컨트롤을 해야할 것들이 추가되면,

MainActivity class 에 너무 많이 책임이 몰릴것 같습니다.

저는 각 Fragment에서 어떤 노출되어야 하는지를 인터페이스화하여 반환만하고 MainActivity는 그 인터페이스에 맞게 동작하는 방식이 좋아보이지만, Navigation 때문에 쉽지 않아보입니다. 😢 

그래도 이런식의 하나의 class에 책임이 몰리는 구조가 좋지 않은 구조라는 것을 인지하셨으면 하는 바램으로 코멘트 남겨봅니다.

cherrylime69 replied:
앨런이 말한 부분을 계속 생각하여 추후에 해당 코드를 변경할 수 있는 방향으로 리팩터링 해보도록 노력하겠습니다. 좋은 피드백 감사합니다! 😊

kang-proto committed:
fix: 첫번째주 수요일 PR 리뷰 반영
- icon 변경 Switch로직 Util로 분리
- visibility-> isVisible 변경
- CardView elevation 설정
- Manifest의 windowSoftInputMode 설정
- 불필요한 라이브러리, 코드 제거, 코드 포맷팅 진행
- Destination 모델의 property 수정

cherrylime69 committed:
Merge pull request #18 from hwicode/fix/firstWeekFirstReview

[Android][히데, 라이너스]  첫번째주 수요일 PR 리뷰 반영

HyungMinKang committed:
Merge pull request #19 from hwicode/android

[Android] [히데&라이너스]  첫번째주 수요일 PR 반영

HyungMinKang committed:
Merge branch 'team-19' into android-develop
